### PR TITLE
[LOOP-1125] Upload pump data

### PR DESCRIPTION
**Pull request by darinkrauss**
- https://tidepool.atlassian.net/browse/LOOP-1125
- https://tidepool.atlassian.net/browse/LOOP-1124
- https://tidepool.atlassian.net/browse/LOOP-3389
- https://tidepool.atlassian.net/browse/LOOP-3388
- Allow scheduled basal rate on basal DoseEntry
- Add duringSuspend property to DoseEntry
- Persist duringSuspend property in CachedInsulinDeliveryObject
- Allow temporary basal during suspend to have scheduled basal rate
- Extend NewPumpEvent and DoseEntry to support RawRepresentable
- Update InsulinDeliveryStore to cache 90 days of date
- Add PumpAlarmType for alarm PumpEvents
- Add SyncPumpEvent to support RemoteDataService
- Combine dose and pump event data for RemoteDataService
- Update DoseStore.executePumpQuery to merge pump events and doses
- Add limit and inclusive to InsulinDeliveryStore.getDoseEntries
- Update model and model mapping
- Update MockPumpManager to allow non-dose pump events
- Update MockPumpManager settings to store pump alarms
- Add and update tests

Note: Needs review from @ps2.

### CODE REVIEW

**Comment by darinkrauss** [comment]
LW PR for build: https://github.com/tidepool-org/LoopWorkspace/pull/741

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift [hunk 1]
```diff
@@ -57,6 +60,9 @@ extension HKQuantitySample {
             if dose.type == .tempBasal {
                 metadata[MetadataKeyProgrammedTempBasalRate] = HKQuantity(unit: .internationalUnitsPerHour, doubleValue: dose.unitsPerHour)
             }
+            if dose.type == .suspend {
+                metadata[MetadataKeyDuringSuspend] = true
+            }
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: Should this be:

```suggestion
            metadata[MetadataKeyDuringSuspend] = dose.type == .suspend
```

if not, why not? Could `duringSuspend` below just be non-optional, like `manuallyEntered` is?  That would avoid awkward dealings with the optional bool.

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift [hunk 2]
```diff
@@ -110,6 +116,10 @@ extension HKQuantitySample {
         return metadata?[MetadataKeyProgrammedTempBasalRate] as? HKQuantity
     }
 
+    var duringSuspend: Bool? {
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: [nit] Consider making this non-optional (see above)

File: LoopKit/InsulinKit/PumpEvent+CoreDataProperties.swift [hunk 3]
```diff
@@ -48,8 +49,10 @@ extension PumpEvent: Encodable {
         try container.encode(mutable, forKey: .mutable)
         try container.encodeIfPresent(raw?.base64EncodedString(), forKey: .raw)
         try container.encodeIfPresent(title, forKey: .title)
+        try container.encodeIfPresent(insulinType, forKey: .insulinType)
+        try container.encodeIfPresent(automatic, forKey: .automatic)
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/PumpEvent+CoreDataProperties.swift:53]: interesting...`automatic` was missing!  Good catch.

File: LoopKitTests/DoseStoreTests.swift [hunk 4]
```diff
@@ -593,194 +570,1042 @@ class DoseStoreQueryTests: PersistenceControllerTestCase {
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        wait(for: [completion], timeout: 10)
+    }
+
+    func testPagingOufOfOrderPumpEvents() {
```
Comment (rickpasetto) [inline LoopKitTests/DoseStoreTests.swift]: [tiny nit] spelling
```suggestion
    func testPagingOutOfOrderPumpEvents() {
```

File: LoopKit/InsulinKit/DoseStore.swift [hunk 5]
```diff
@@ -1646,49 +1627,121 @@ extension DoseStore {
...
+            switch result {
+            case .failure(let error):
+                completion(.failure(error))
+            case .success(let doses):
+                var doseEvents = doses.compactMap { $0.syncPumpEventWithModificationCounter }
+                doseEvents.sort(by: { SyncPumpEventWithModificationCounter.sort($0, $1) })
+
+                var events: [SyncPumpEventWithModificationCounter] = []
+
+                // Walk the two sets of events and reconcile. As the events are sorted by start date, the event with the earlier
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/DoseStore.swift]: [nit] Consider splitting this logic out into separate function[s].  As it is, it is kinda hard to follow and get the big picture, all inside this one `executePumpQuery` function. The comments are helpful, though, thanks.

**Review by rickpasetto — APPROVED** [review]
Mostly nits.  It'd be really great to see the long `executePumpQuery` function refactored into smaller functions, but not blocking approval.

File: LoopKit/InsulinKit/CachedInsulinDeliveryObject+CoreDataClass.swift [hunk 6]
```diff
@@ -111,6 +111,19 @@ class CachedInsulinDeliveryObject: NSManagedObject {
             primitiveAutomaticallyIssued = newValue != nil ? NSNumber(booleanLiteral: newValue!) : nil
         }
     }
+
+    var duringSuspend: Bool? {
```
Comment (ps2) [inline LoopKit/InsulinKit/CachedInsulinDeliveryObject+CoreDataClass.swift]: As discussed in Slack DMs...

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift [hunk 7]
```diff
@@ -166,23 +166,24 @@ extension InsulinDeliveryStore {
     /// - Parameters:
     ///   - start: The earliest date of dose entries to retrieve, if provided.
     ///   - end: The latest date of dose entries to retrieve, if provided.
+    ///   - limit: The maximum number of dose entries to retrieve, if provided, otherwise unlimited.
     ///   - completion: A closure called once the dose entries have been retrieved.
     ///   - result: An array of dose entries, in chronological order by startDate, or error.
-    public func getDoseEntries(start: Date? = nil, end: Date? = nil, completion: @escaping (_ result: Result<[DoseEntry], Error>) -> Void) {
+    public func getDoseEntries(start: Date? = nil, end: Date? = nil, limit: Int? = nil, inclusive: Bool = true, completion: @escaping (_ result: Result<[DoseEntry], Error>) -> Void) {
```
Comment (ps2) [inline LoopKit/InsulinKit/InsulinDeliveryStore.swift]: Is this new inclusive flag used just for the test, or is it planned for some other client use?

File: LoopKit/InsulinKit/PumpAlarmType.swift [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+public enum PumpAlarmType: Equatable, Codable {
+    case autoOff
+    case lowInsulin
+    case lowPower
+    case noDelivery
+    case noInsulin
+    case noPower
+    case occlusion
+    case other(_ details: String)
+    case overLimit
```
Comment (ps2) [inline LoopKit/InsulinKit/PumpAlarmType.swift]: This (`overLimit`) seems pretty vague. Is there are particular kind of limit that this is intended for?

Comment (ps2) [reply LoopKit/InsulinKit/PumpAlarmType.swift]: Maybe it's for expiration?  Having a pump expired alarm seems pretty useful. Would recommend changing the name in that case to be more explicit.

File: LoopKit/InsulinKit/SyncPumpEvent.swift [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+//  SyncPumpEvent.swift
+//  LoopKit
+//
+//  Created by Darin Krauss on 12/16/21.
+//  Copyright © 2021 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+
+public struct SyncPumpEvent: Equatable {
```
Comment (ps2) [inline LoopKit/InsulinKit/SyncPumpEvent.swift]: `PersistedPumpEvent` is the type in LoopKit that is designed for representing pump events stored in DoseStore. Why is this being created and returned from queries instead?

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift (hunk 1 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: No longer necessary. Removing `duringSuspend` per other PR feedback.

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift (hunk 2 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: No longer necessary. Removing `duringSuspend` per other PR feedback.

File: LoopKit/InsulinKit/SyncPumpEvent.swift (hunk 9 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/SyncPumpEvent.swift]: Primarily because `PersistedPumpEvent` has a number of fields that don't need to be exposed to the remote data services. Plus, `type` is optional while it should really be required for remote data services. It isn't a big deal, but if you feel strongly I can replace it.

File: LoopKit/InsulinKit/PumpAlarmType.swift (hunk 8 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/PumpAlarmType.swift]: I'll remove it.

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift (hunk 7 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/InsulinDeliveryStore.swift]: It is required by the DoseStore when fetching doses for reconciling with pump events for upload to remote data services. It defaults to `true`, but in this case it must be set to `false`.

File: LoopKit/InsulinKit/CachedInsulinDeliveryObject+CoreDataClass.swift (hunk 6 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/CachedInsulinDeliveryObject+CoreDataClass.swift]: Renamed to `isSuspend` and made non-optional. Removed from `DoseEntry`.

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift (hunk 2 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: Actually, necessary here, but not in DoseEntry. Made non-optional per feedback.

File: LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift (hunk 1 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/HKQuantitySample+InsulinKit.swift]: Actually, necessary here, but not in DoseEntry. Made non-optional per feedback.

**Comment by darinkrauss** [comment]
@ps2 @rickpasetto Ready for another review.

File: LoopKit/InsulinKit/DoseStore.swift [hunk 10]
```diff
@@ -1640,86 +1640,8 @@ extension DoseStore {
...
-                        // the remote likelyhood of this last case.
-                        if let limitIndex = limitEvents.firstIndex(where: { $0.modificationCounter.map { $0 > extraModificationCounterEarliest } == true }), limitIndex > 0 {
-                            (events, _) = limitEvents.split(at: limitIndex)
-                        }
-                    } else {
-                        events = limitEvents
-                    }
-                }
+                // Reconcile and limit the events
+                let events = self.limitPumpEvents(self.reconcilePumpEventsAndDoseEvents(pumpEvents: pumpEvents, doseEvents: doseEvents), limit: limit)
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/DoseStore.swift]: 👍

File: LoopKit/InsulinKit/DoseStore.swift [hunk 11]
```diff
@@ -1737,11 +1659,104 @@ extension DoseStore {
         }
     }
 
-    fileprivate static var storedDateFormatter: ISO8601DateFormatter = {
-        let dateFormatter = ISO8601DateFormatter()
-        dateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
-        return dateFormatter
-    }()
+    // Walk the two sets of events and reconcile based upon start date and type.
+    private func reconcilePumpEventsAndDoseEvents(pumpEvents: [SyncPumpEventWithModificationCounter], doseEvents: [SyncPumpEventWithModificationCounter]) -> [SyncPumpEventWithModificationCounter] {
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/DoseStore.swift]: 👍 And I like how you split the responsibilities with clear function params.

**Review by rickpasetto — APPROVED** [review]
LGTM (thanks for refactoring those functions)

File: LoopKit/InsulinKit/SyncPumpEvent.swift (hunk 9 above)
Comment (ps2) [reply LoopKit/InsulinKit/SyncPumpEvent.swift]: Exposing a protocol is a nice way to restrict visibility if you really need to hide internals from a public api. You can make PersistedPumpEvent As for `type`, what happens if you encounter a pump event where it is nil? What does that look like when being turned into a `SyncPumpEvent`? It does seem like the sole role of PersistedPumpEvent was exposing pump events to the query interface, so I think now we have two types that users of DoseStore will see that represent PumpEvents stored in DoseStore.

Comment (darinkrauss) [reply LoopKit/InsulinKit/SyncPumpEvent.swift]: Specifically, what do you want me to do here.

Comment (darinkrauss) [reply LoopKit/InsulinKit/SyncPumpEvent.swift]: Regarding your questions about when then pump event type is `nil`, it is explicitly ignored. The fact that the `type` property is optional appears to be an artifact of history. The `type` of a pump event will either be the explicitly specified type or the type of the explicitly specified dose. While it is technically possible from an API perspective to specify `nil` for both, it is not in actual usage (except in some tests). Therefore, it is an "invalid" pump event and, thus, ignored.

**Review by ps2 — COMMENTED** [review]
Just marking as reviewed, so I can get notified when the next iteration is ready that we discussed.

**Comment by darinkrauss** [comment]
@ps2 @rickpasetto Ready for another review. Hopefully this is the last one.

**Review by rickpasetto — COMMENTED** [review]
These new changes look ok to me, but I defer to @ps2 for final say-so.

File: LoopKit/InsulinKit/DoseStore.swift [hunk 12]
```diff
@@ -846,19 +836,19 @@ extension DoseStore {
...
+     Adds and persists doses. Doses *cannot* be mutable.
      - parameter doses: An array of dose entries to add.
      - parameter completion: A closure called after the doses are saved. The closure takes a single argument:
      - parameter error: An error object explaining why the doses could not be saved.
      */
-    public func addDoses(_ doses: [DoseEntry], completion: @escaping (_ error: Error?) -> Void) {
-        guard doses.count > 0 else {
+    public func addDoses(_ doses: [DoseEntry], from device: HKDevice?, completion: @escaping (_ error: Error?) -> Void) {
+        guard doses.count > 0, !doses.contains(where: { $0.isMutable }) else {
             completion(nil)
```
Comment (ps2) [inline LoopKit/InsulinKit/DoseStore.swift:847]: Is it a programming error to call `addDoses` with a mutable dose? If so, should this be an assertion?

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift [hunk 13]
```diff
@@ -142,17 +156,22 @@ public class InsulinDeliveryStore: HealthKitSampleStore {
...
 
-            if !changed {
+            guard dosesAdded || dosesDeleted else {
                 completion(true)
                 return
             }
 
             self.handleUpdatedDoseData(updateSource: .queriedByHealthKit)
+
+            if dosesAdded {
```
Comment (ps2) [inline LoopKit/InsulinKit/InsulinDeliveryStore.swift]: Do we not want to notify if doses were deleted? If so, why not?

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift [hunk 14]
```diff
@@ -273,7 +335,7 @@ extension InsulinDeliveryStore {
     ///   - syncVersion: The sync version used for the new dose entries.
     ///   - completion: A closure called once the dose entries have been stored.
     ///   - result: Success or error.
-    func addDoseEntries(_ entries: [DoseEntry], from device: HKDevice?, syncVersion: Int, manuallyEntered: Bool = false, completion: @escaping (_ result: Result<Void, Error>) -> Void) {
+    func addDoseEntries(_ entries: [DoseEntry], from device: HKDevice?, syncVersion: Int, resolveMutable: Bool = false, completion: @escaping (_ result: Result<Void, Error>) -> Void) {
```
Comment (ps2) [inline LoopKit/InsulinKit/InsulinDeliveryStore.swift:339]: Could you update the documentation to describe what use the `resolveMutable` parameter is for?

Comment (ps2) [reply LoopKit/InsulinKit/InsulinDeliveryStore.swift:339]: And is passing false (or using the default) necessary for some cases, or is it just an optimization for those times we don't need it?

**Review by ps2 — APPROVED** [review]
LGTM. Mainly some questions. 

Also, does this still migrate from DIY master? I see the mapping was updated, so I'm guessing it does. And does it auto-migrate from the previous version?

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift (hunk 13 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/InsulinDeliveryStore.swift]: Sure, we can do that.

File: LoopKit/InsulinKit/InsulinDeliveryStore.swift (hunk 14 above)
Comment (darinkrauss) [reply LoopKit/InsulinKit/InsulinDeliveryStore.swift:339]: This is just an optimization for when we don't need it. We don't want to resolve mutable dose entries *except* when we know the incoming dose entries will properly update any pre-existing mutable dose entries. This is only the case when called from `DoseStore.addPumpEvents`.
